default/python/AI/PolicyAI.py

```diff
@@ -379,29 +381,108 @@ def _rate_algo_research(self) -> float:
         debug(f"_rate_algo_research: rating={rating}")
         return rating
 
-    def _process_liberty(self) -> None:
+    def _conformance_or_liberty(self) -> _EmpireOutput:
         """
-        Adopt or deadopt liberty, may replace another social policy.
+        Try policies conformance and liberty.
         Liberty is available very early, but to help setting up basic policies we keep propaganda at least
         until we have infra1. Getting a second slot before that should be nearly impossible.
-        Centralization is always kept only for one turn and if population has a strong opinion on it,
-        stability calculation may be extremely different from normal turns, leading to a change that would
-        possibly be reverted next turn, so we skip processing liberty while centralization is adopted.
+        """
+        if infra1 in self._adopted:
+            debug("Evaluating conformance/liberty, first without change:")
+            current_output = self._calculate_empire_output()
+            if conformance in self._adopted:
+                # this may also try liberty
+                self._try_without_conformance(current_output)
+            else:
+                current_output, conformance_cost = self._try_conformance(current_output)
+                self._process_liberty(current_output, conformance_cost)
+
+    def _try_without_conformance(self, current_output: _EmpireOutput) -> None:
```

Comment:
The thing that bothers me, it that this is a huge function with toons of variables, ifs, and loops. 

maybe move it to a separate file and try to split it into small chunks.

Reply:
Hey, it wasn't that complex.

Perhaps you like it better now?